### PR TITLE
Update Seldon to resulve issues with k8s 1.16/OCP 4.3

### LINKED ISSUE #14: Seldon CRD fails to deploy on OpenShift 4.3
OpenShift 4.3 comes with an update to Kubernetes 1.16 which results in some incompatibilities - this time for Seldon CRD to fail to be applied.

It is already known to Kubeflow community - https://github.com/kubeflow/manifests/issues/375

@cliveseldon I see your changes in upstream basically resolve to replacing the resources with new version in master, what would you suggest to do here as we would like to keep 0.7 working on OCP 4.3 (i.e. k8s 1.16)?

### PR DESCRIPTION
**Pull request by vpavlin**
**Which issue is resolved by this Pull Request:**
Resolves #14 

@cliveseldon PTAL

**Description of your changes:**

Cherry picked Seldon update (https://github.com/kubeflow/manifests/commit/c4f51b1cc0b4854b6f6180d5a3bb021dd65691a8#diff-56fba03a2d55cd9b61df7e387d1d6d47) from manifests master and tested it on OCP 4.3

I also had to bump memory limit for seldon from 30Mi to 300Mi - it was getting OOM killed and add `cert-manager` to `kfctl_openshfit.yaml`. I also had to change the webhook port to `8443`

**Steps to verify:**

```
sed -i 's#uri: .*#uri: '$PWD'#' ./kfdef/kfctl_openshift.yaml
rm -rf kfdef/.cache kfdef/kustomize/ && kfctl build -V -f kfdef/kfctl_openshift.yaml 
kfctl apply -V -f kfdef/kfctl_openshift.yaml
```

You should see all pods in `istio-system`, `cert-manager` and `kubeflow` namespaces running without issue

### CODE REVIEW

File: seldon/seldon-core-operator/base/resources.yaml [hunk 1]
```diff
@@ -0,0 +1,3682 @@
...
+                                          properties:
+                                            name:
+                                              description: 'Name of the referent. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names TODO: Add other useful fields. apiVersion, kind, uid?'
+                                              type: string
+                                            optional:
+                                              description: Specify whether the ConfigMap must be defined
+                                              type: boolean
+                                          type: object
+                                        prefix:
+
```
Comment (vpavlin) [inline seldon/seldon-core-operator/base/resources.yaml]: Ah, need to remove this

**Comment by vpavlin** [comment]
Actually, I get errors

```
WARN[0219] Encountered error during apply:  (kubeflow.error): Code 500 with message: Apply.Run  Error [error when creating "/tmp/kout142091474": Internal error occurred: failed calling webhook "webhook.cert-manager.io": the server is currently unable to handle the request, error when creating "/tmp/kout142091474": Internal error occurred: failed calling webhook "webhook.cert-manager.io": the server is currently unable to handle the request]  filename="kustomize/kustomize.go:183"
```

But if I keep `kfctl` retry couple times, it starts working in the end...@cliveseldon any idea what might be causing this?

**Comment by vpavlin** [comment]
Ah, maybe these errors in cert-manager:

```
I0212 16:17:26.562916       1 sources.go:94] cert-manager/inject-controller "level"=0 "msg"="Extracting CA from Certificate resource" "resource_kind"="CustomResourceDefinition" "resource_name"="seldondeployments.machinelearning.seldon.io" "resource_namespace"="" "certificate"="kubeflow/seldon-serving-cert"
E0212 16:17:26.562971       1 sources.go:110] cert-manager/inject-controller "msg"="unable to fetch associated certificate" "error"="Certificate.cert-manager.io \"seldon-serving-cert\" not found" "certificate"={"Namespace":"kubeflow","Name":"seldon-serving-cert"} "resource_kind"="CustomResourceDefinition" "resource_name"="seldondeployments.machinelearning.seldon.io" "resource_namespace"="" 
I0212 16:17:26.563003       1 controller.go:158] cert-manager/inject-controller "level"=0 "msg"="could not find any ca data in data source for target" "resource_kind"="CustomResourceDefinition" "resource_name"="seldondeployments.machinelearning.seldon.io" "resource_namespace"="" 
```

**Comment by vpavlin** [comment]
Ok, so I was able deploy a model using this following an example in our blog post (https://developers.redhat.com/blog/2020/02/10/installing-kubeflow-v0-7-on-openshift-4-2/)

The issue during deployment is still there and I was able to trace it back to cert-manager taking a lot of time to generate secrets - the errors go back to cert-manager-webhook pod failing to start with

```
MountVolume.SetUp failed for volume "certs" : secret "cert-manager-webhook-tls" not found
```

for couple minutes, once it comes up the installation finishes and then it takes couple more minutes to start Seldon with the same error (failed for volume..)

@cliveseldon Have you seen this before with KF on vanilla Kubernetes?

**Comment by vpavlin** [comment]
Just to follow up on my previous comments - 

* the change works for me
* kfctl 0.7 needs couple retries
* kfctl 1.0.RC also needs to retry, but seems like needs less of them:)

@crobby I think it is ready for review

**Comment by crobby** [comment]
My first try for this (on CRC 1.6.0, which is OpenShift 4.3.0) give me a seldon-operator-controller-manager pod that dies with a log message of:
```
{"level":"error","ts":1582143163.1843057,"logger":"kubebuilder.source","msg":"if kind is a CRD, it should be installed before calling Start","kind":{"Group":"machinelearning.seldon.io","Kind":"SeldonDeployment"},"error":"no matches for kind \"SeldonDeployment\" in version \"machinelearning.seldon.io/v1alpha2\"","stacktrace":"github.com/seldonio/seldon-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/src/github.com/seldonio/seldon-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/seldonio/seldon-operator/vendor/sigs.k8s.io/controller-runtime/pkg/source.(*Kind).Start\n\t/go/src/github.com/seldonio/seldon-operator/vendor/sigs.k8s.io/controller-runtime/pkg/source/source.go:89\ngithub.com/seldonio/seldon-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Watch\n\t/go/src/github.com/seldonio/seldon-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:122\ngithub.com/seldonio/seldon-operator/pkg/controller/seldondeployment.add\n\t/go/src/github.com/seldonio/seldon-operator/pkg/controller/seldondeployment/seldondeployment_controller.go:85\ngithub.com/seldonio/seldon-operator/pkg/controller/seldondeployment.Add\n\t/go/src/github.com/seldonio/seldon-operator/pkg/controller/seldondeployment/seldondeployment_controller.go:68\ngithub.com/seldonio/seldon-operator/pkg/controller.AddToManager\n\t/go/src/github.com/seldonio/seldon-operator/pkg/controller/controller.go:29\nmain.main\n\t/go/src/github.com/seldonio/seldon-operator/cmd/manager/main.go:75\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:198"}
{"level":"error","ts":1582143163.184392,"logger":"entrypoint","msg":"unable to register controllers to the manager","error":"no matches for kind \"SeldonDeployment\" in version \"machinelearning.seldon.io/v1alpha2\"","stacktrace":"github.com/seldonio/seldon-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/src/github.com/seldonio/seldon-operator/vendor/github.com/go-logr/zapr/zapr.go:128\nmain.main\n\t/go/src/github.com/seldonio/seldon-operator/cmd/manager/main.go:76\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:198"}

```

It's not even seeing the SeldonDeployment CRD.  I will try again to make sure it wasn't anything goofy that I did.

**Comment by crobby** [comment]
> My first try for this (on CRC 1.6.0, which is OpenShift 4.3.0) give me a seldon-operator-controller-manager pod that dies with a log message of:
> 
> ```
> {"level":"error","ts":1582143163.1843057,"logger":"kubebuilder.source","msg":"if kind is a CRD, it should be installed before calling Start","kind":{"Group":"machinelearning.seldon.io","Kind":"SeldonDeployment"},"error":"no matches for kind \"SeldonDeployment\" in version \"machinelearning.seldon.io/v1alpha2\"","stacktrace":"github.com/seldonio/seldon-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/src/github.com/seldonio/seldon-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/seldonio/seldon-operator/vendor/sigs.k8s.io/controller-runtime/pkg/source.(*Kind).Start\n\t/go/src/github.com/seldonio/seldon-operator/vendor/sigs.k8s.io/controller-runtime/pkg/source/source.go:89\ngithub.com/seldonio/seldon-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Watch\n\t/go/src/github.com/seldonio/seldon-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:122\ngithub.com/seldonio/seldon-operator/pkg/controller/seldondeployment.add\n\t/go/src/github.com/seldonio/seldon-operator/pkg/controller/seldondeployment/seldondeployment_controller.go:85\ngithub.com/seldonio/seldon-operator/pkg/controller/seldondeployment.Add\n\t/go/src/github.com/seldonio/seldon-operator/pkg/controller/seldondeployment/seldondeployment_controller.go:68\ngithub.com/seldonio/seldon-operator/pkg/controller.AddToManager\n\t/go/src/github.com/seldonio/seldon-operator/pkg/controller/controller.go:29\nmain.main\n\t/go/src/github.com/seldonio/seldon-operator/cmd/manager/main.go:75\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:198"}
> {"level":"error","ts":1582143163.184392,"logger":"entrypoint","msg":"unable to register controllers to the manager","error":"no matches for kind \"SeldonDeployment\" in version \"machinelearning.seldon.io/v1alpha2\"","stacktrace":"github.com/seldonio/seldon-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/src/github.com/seldonio/seldon-operator/vendor/github.com/go-logr/zapr/zapr.go:128\nmain.main\n\t/go/src/github.com/seldonio/seldon-operator/cmd/manager/main.go:76\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:198"}
> ```
> 
> It's not even seeing the SeldonDeployment CRD. I will try again to make sure it wasn't anything goofy that I did.

Ok, it looks like I was just impatient in waiting for the webhook cert issue to self-resolve.  the kfctl apply operation is now succeeding for me and I *do* see the CRD installed, but I'm still seeing the following for my seldon operator controller pod:

`{"level":"error","ts":1582143825.28458,"logger":"entrypoint","msg":"unable to run the manager","error":"open /tmp/cert/cert.pem: no such file or directory","stacktrace":"github.com/seldonio/seldon-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/src/github.com/seldonio/seldon-operator/vendor/github.com/go-logr/zapr/zapr.go:128\nmain.main\n\t/go/src/github.com/seldonio/seldon-operator/cmd/manager/main.go:89\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:198"}
`

**Comment by crobby** [comment]
> > My first try for this (on CRC 1.6.0, which is OpenShift 4.3.0) give me a seldon-operator-controller-manager pod that dies with a log message of:
> > ```
> > {"level":"error","ts":1582143163.1843057,"logger":"kubebuilder.source","msg":"if kind is a CRD, it should be installed before calling Start","kind":{"Group":"machinelearning.seldon.io","Kind":"SeldonDeployment"},"error":"no matches for kind \"SeldonDeployment\" in version \"machinelearning.seldon.io/v1alpha2\"","stacktrace":"github.com/seldonio/seldon-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/src/github.com/seldonio/seldon-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/seldonio/seldon-operator/vendor/sigs.k8s.io/controller-runtime/pkg/source.(*Kind).Start\n\t/go/src/github.com/seldonio/seldon-operator/vendor/sigs.k8s.io/controller-runtime/pkg/source/source.go:89\ngithub.com/seldonio/seldon-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Watch\n\t/go/src/github.com/seldonio/seldon-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:122\ngithub.com/seldonio/seldon-operator/pkg/controller/seldondeployment.add\n\t/go/src/github.com/seldonio/seldon-operator/pkg/controller/seldondeployment/seldondeployment_controller.go:85\ngithub.com/seldonio/seldon-operator/pkg/controller/seldondeployment.Add\n\t/go/src/github.com/seldonio/seldon-operator/pkg/controller/seldondeployment/seldondeployment_controller.go:68\ngithub.com/seldonio/seldon-operator/pkg/controller.AddToManager\n\t/go/src/github.com/seldonio/seldon-operator/pkg/controller/controller.go:29\nmain.main\n\t/go/src/github.com/seldonio/seldon-operator/cmd/manager/main.go:75\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:198"}
> > {"level":"error","ts":1582143163.184392,"logger":"entrypoint","msg":"unable to register controllers to the manager","error":"no matches for kind \"SeldonDeployment\" in version \"machinelearning.seldon.io/v1alpha2\"","stacktrace":"github.com/seldonio/seldon-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/src/github.com/seldonio/seldon-operator/vendor/github.com/go-logr/zapr/zapr.go:128\nmain.main\n\t/go/src/github.com/seldonio/seldon-operator/cmd/manager/main.go:76\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:198"}
> > ```
> > 
> > 
> > It's not even seeing the SeldonDeployment CRD. I will try again to make sure it wasn't anything goofy that I did.
> 
> Ok, it looks like I was just impatient in waiting for the webhook cert issue to self-resolve. the kfctl apply operation is now succeeding for me and I _do_ see the CRD installed, but I'm still seeing the following for my seldon operator controller pod:
> 
> `{"level":"error","ts":1582143825.28458,"logger":"entrypoint","msg":"unable to run the manager","error":"open /tmp/cert/cert.pem: no such file or directory","stacktrace":"github.com/seldonio/seldon-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/src/github.com/seldonio/seldon-operator/vendor/github.com/go-logr/zapr/zapr.go:128\nmain.main\n\t/go/src/github.com/seldonio/seldon-operator/cmd/manager/main.go:89\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:198"} `

Update:  waiting a bit longer seems to give me a seldon operator controller manager pod that isn't erroring-out, but when I try to create a SeldonDeployment I get:

`oc create -f testseldon.yaml`
`Error from server (InternalError): error when creating "testseldon.yaml": Internal error occurred: failed calling webhook "mutating-create-update-seldondeployment.seldon.io": Post https://webhook-server-service.kubeflow.svc:443/mutating-create-update-seldondeployment?timeout=30s: dial tcp 10.128.0.103:9876: connect: connection refused
`

**Comment by crobby** [comment]
> failed calling webhook "mutating-create-update-seldondeployment.seldon.io"

And once again, it seems like I was just too impatient.  I let my install "settle" for 15-20 min and when I came back, I tried creating my SeldonDeployment again.  To my surprise, it worked this time.

I do have concerns that its taking this long for the install to become stable/usable, but it is at least working, which is an improvement over the current behavior on 4.3.  Does this also work on 4.2, or has that not yet been tested?

**Comment by crobby** [comment]
When using my CRC instance, I am seeing my seldoncontrollermanager pod get killed periodically with the message:  

`The node was low on resource: ephemeral-storage. Container manager was using 964Ki, which exceeds its request of 0. `

I'm not yet sure if this is a problem specific to CRC or if it could be a more general problem.

**Comment by vpavlin** [comment]
I will give it one more go on bigger cluster before merge to see if I hit any of your issues - I do not remember seeing them, so it really might be just flakiness of CRC.

Thanks!

**Comment by crobby** [comment]
> I will give it one more go on bigger cluster before merge to see if I hit any of your issues - I do not remember seeing them, so it really might be just flakiness of CRC.
> 
> Thanks!

I discovered that CRC (as of 1.6.0) has changed the disk structure of the VM, which I believe caused my cluster to less disk than I thought.  After I corrected that this morning, my install seems to be much more stable.  The only bit of flake is at kfctl apply time:

`WARN[0047] Encountered error during apply:  (kubeflow.error): Code 500 with message: Apply.Run  Error error when creating "/tmp/kout398360272": Internal error occurred: failed calling webhook "webhook.cert-manager.io": the server is currently unable to handle the request  filename="kustomize/kustomize.go:183"`

But it does go away after a minute or two.

**Comment by vpavlin** [comment]
Yes, I am also seeing the cert webhook errors, but I think that is "expected" since Seldon needs the cert-manager to start and there is no "wait for" option in Kubeflow, so it just have to retry.

**Comment by crobby** [comment]
fwiw:  I also tried this on OpenShift 4.2.9 and everything seemed happy there too.

**Pull request merged**